vaporetto/src/predictor.rs

```diff
@@ -1,1136 +1,861 @@
+            #[cfg(feature = "tag-prediction")]
+            tag_char_ngram_model,
+        )?;
+        let type_scorer = TypeScorer::new(
+            model.0.type_ngram_model,
+            model.0.type_window_size,
+            #[cfg(feature = "tag-prediction")]
+            tag_type_ngram_model,
+        )?;
+        Ok(Self(PredictorData {
+            char_scorer,
+            type_scorer,
+            bias: model.0.bias,
+
+            #[cfg(feature = "tag-prediction")]
+            tag_predictor,
+
+            #[cfg(feature = "tag-prediction")]
+            n_tags,
+        }))
+    }
+
+    /// Predicts word boundaries of the given sentence.
+    /// If necessary, this function also prepares for predicting tags.
+    pub fn predict<'a, 'b>(&'b self, sentence: &mut Sentence<'a, 'b>) {
+        sentence.score_padding = WEIGHT_FIXED_LEN - 1;
+        sentence.boundary_scores.clear();
+        sentence
+            .boundary_scores
+            .resize(sentence.score_padding * 2 + sentence.len() - 1, self.0.bias);
+        if let Some(scorer) = self.0.char_scorer.as_ref() {
+            scorer.add_scores(sentence);
+        }
+        if let Some(scorer) = self.0.type_scorer.as_ref() {
+            scorer.add_scores(sentence);
         }
-        let mut right_scores_iter = sentence.tag_scores.right_scores.chunks(n_tags);
-        let mut last_boundary_idx = 0;
-        for (i, ((((b, left_scores), right_scores), self_scores), tag)) in sentence
+        for (b, s) in sentence
             .boundaries
-            .iter()
-            .zip(left_scores_iter)
-            .zip(&mut right_scores_iter)
-            .zip(&sentence.tag_scores.self_scores)
-            .zip(&mut sentence.tags)
-            .enumerate()
+            .iter_mut()
+            .zip(&sentence.boundary_scores[sentence.score_padding..])
         {
-            if *b == BoundaryType::WordBoundary {
-                for (t, r) in tag_score.iter_mut().zip(right_scores) {
-                    *t += *r;
-                }
-                if let Some(self_weights) = self_scores.as_ref() {
-                    let diff = -i16::try_from(i + 1 - last_boundary_idx).unwrap_or(0);
-                    for self_weight in self_weights.iter() {
-                        match self_weight.start_rel_position.cmp(&diff) {
-                            Ordering::Greater => continue,
-                            Ordering::Equal => {
-                                for (t, s) in tag_score.iter_mut().zip(&self_weight.weight) {
-                                    *t += *s;
-                                }
+            if *s > 0 {
+                *b = CharacterBoundary::WordBoundary;
+            } else {
+                *b = CharacterBoundary::NotWordBoundary;
+            }
+        }
+        sentence.set_predictor(self);
+    }
+
+    #[cfg(feature = "tag-prediction")]
+    pub(crate) fn predict_tags<'a, 'b>(&'b self, sentence: &mut Sentence<'a, 'b>) {
+        if self.0.n_tags == 0 {
```

Comment:
It is necessary to check if `self.tag_predictor = Some` or `None`, following the above modification.

Reply:
d2acb55